File: framework/data-jpa/src/test/java/cn/bestwu/simpleframework/data/test/test/SoftDeleteTest.java
package cn.bestwu.simpleframework.data.test.test;

import cn.bestwu.simpleframework.data.test.domain.HardUser;
import cn.bestwu.simpleframework.data.test.domain.QUser;
import cn.bestwu.simpleframework.data.test.domain.User;
import cn.bestwu.simpleframework.data.test.repository.HardUserRepository;
import cn.bestwu.simpleframework.data.test.repository.UserRepository;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import javax.sql.DataSource;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringRunner;

@RunWith(SpringRunner.class)
@SpringBootTest
public class SoftDeleteTest {

  @Autowired
  UserRepository repository;
  @Autowired
  HardUserRepository hardUserRepository;
  @Autowired
  DataSource dataSource;
  final List<User> batch = new ArrayList<>();
  final List<Integer> batchIds = new ArrayList<>();
  Integer daveId;
  Integer carterId;

  @Before
  public void setUp() {
//    RunScript.execute(dataSource.getConnection(),
//        new FileReader(new ClassPathResource("import.sql").getFile()));

    User dave = new User("Dave", "Matthews");
    repository.save(dave);
    User dave1 = new User("Dave", "Matthews");
    repository.save(dave1);
    User carter = new User("Carter", "Beauford");
    repository.save(carter);
    carter = new User("Carter", "Beauford");
    repository.save(carter);

    Collections.addAll(batch, dave, dave1);
    daveId = dave.getId();
    Collections.addAll(batchIds, daveId, dave1.getId());

    repository.delete(dave);
    carterId = carter.getId();

  }

  @Test
  public void name() {
    System.err.println(repository.findAll(QUser.user.firstname.contains("D")));
  }

  @After
  public void tearDown() {
    repository.deleteAll();
    repository.cleanRecycleBin();
    hardUserRepository.deleteAll();
    hardUserRepository.cleanRecycleBin();
  }

  @Test
  public void hardDeleteTest() {
    HardUser dave = new HardUser("Dave", "Matthews");
    hardUserRepository.save(dave);
    hardUserRepository.delete(dave);
    Optional<HardUser> optionalUser = hardUserRepository.findByIdFromRecycleBin(dave.getId());
    optionalUser.ifPresent(System.out::println);
    Assert.assertFalse(optionalUser.isPresent());
  }

  @Test
  public void methdQuery() {
    repository.deleteInBatch(batch);

    List<User> users = repository.findByLastname("Matthews");
    System.err.println(users);
    Assert.assertEquals(0, users.size());

    users = repository.findByLastname("Beauford");
    System.err.println(users);
    Assert.assertEquals(2, users.size());

    List<User> recycleAll = repository.findAllFromRecycleBin();
    System.err.println(recycleAll);
    Assert.assertEquals(2, recycleAll.size());

  }

  @Test
  public void methdDelete() {
    repository.deleteByLastname("Matthews");

    List<User> users = repository.findByLastname("Matthews");
    System.err.println(users);
    Assert.assertEquals(0, users.size());

    users = repository.findByLastname("Beauford");
    System.err.println(users);
    Assert.assertEquals(2, users.size());

    List<User> recycleAll = repository.findAllFromRecycleBin();
    System.err.println(recycleAll);
    Assert.assertEquals(2, recycleAll.size());

  }
}

File: data-jpa/src/test/kotlin/top/bettercode/summer/data/jpa/test/SoftDeleteTest.kt
package top.bettercode.summer.data.jpa.test

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.junit.jupiter.SpringExtension
import top.bettercode.summer.data.jpa.domain.HardUser
import top.bettercode.summer.data.jpa.domain.User
import top.bettercode.summer.data.jpa.repository.HardUserRepository
import top.bettercode.summer.data.jpa.repository.UserRepository
import java.util.*
import javax.sql.DataSource

@ExtendWith(SpringExtension::class)
@SpringBootTest
class SoftDeleteTest {
    @Autowired
    lateinit var repository: UserRepository

    @Autowired
    var hardUserRepository: HardUserRepository? = null

    @Autowired
    var dataSource: DataSource? = null
    val batch: MutableList<User> = ArrayList()
    val batchIds: MutableList<Int?> = ArrayList()
    var daveId: Int = 0
    var carterId: Int = 0

    @BeforeEach
    fun setUp() {
        val dave = User("Dave", "Matthews")
        repository.save(dave)
        val dave1 = User("Dave", "Matthews")
        repository.save(dave1)
        var carter = User("Carter", "Beauford")
        repository.save(carter)
        carter = User("Carter", "Beauford")
        repository.save(carter)
        Collections.addAll(batch, dave, dave1)
        daveId = dave.id!!
        Collections.addAll(batchIds, daveId, dave1.id)
        repository.delete(dave)
        carterId = carter.id!!
        System.err.println("--------------------------------------------------------")
    }

    @AfterEach
    fun tearDown() {
        System.err.println("--------------------------------------------------------")
        repository.deleteAll()
        repository.cleanRecycleBin()
        hardUserRepository!!.deleteAll()
        hardUserRepository!!.cleanRecycleBin()
    }

    @Test
    fun hardDeleteTest() {
        val dave = HardUser("Dave", "Matthews")
        hardUserRepository!!.save(dave)
        hardUserRepository!!.delete(dave)
        val optionalUser = hardUserRepository!!.findByIdFromRecycleBin(dave.id)
        optionalUser.ifPresent { x: HardUser? -> println(x) }
        Assertions.assertFalse(optionalUser.isPresent)
    }

    @Test
    fun methdQuery() {
        repository.deleteAllInBatch(batch)
        var users = repository.findByLastName("Matthews")
        System.err.println(users)
        Assertions.assertEquals(0, users!!.size)
        users = repository.findByLastName("Beauford")
        System.err.println(users)
        Assertions.assertEquals(2, users!!.size)
        val recycleAll = repository.findAllFromRecycleBin()
        System.err.println(recycleAll)
        Assertions.assertEquals(2, recycleAll.size)
    }

    @Test
    fun methdDelete() {
        repository.deleteByLastName("Matthews")
        var users = repository.findByLastName("Matthews")
        System.err.println(users)
        Assertions.assertEquals(0, users!!.size)
        users = repository.findByLastName("Beauford")
        System.err.println(users)
        Assertions.assertEquals(2, users!!.size)
        val recycleAll = repository.findAllFromRecycleBin()
        System.err.println(recycleAll)
        Assertions.assertEquals(2, recycleAll.size)
    }
}